File: dc/Overnight_Development_Squad_T2/AirlineReservationSystem/src/ec/edu/espe/airlinereservationsystem/view/MenuManager.java
package ec.edu.espe.airlinereservationsystem.view;

import utils.ReservationSystem;
import enums.TicketClass;
import enums.PaymentMethods;
import ec.edu.espe.airlinereservationsystem.model.Customer;
import ec.edu.espe.airlinereservationsystem.model.Flight;
import ec.edu.espe.airlinereservationsystem.model.Ticket;
import utils.CustomerDataManager;
import utils.JsonUtils;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Scanner;

/**
 *
 * @author Julio Blacio, Overnight Developers Squad, DCCO-ESPE
 */
public class MenuManager {

    private final ReservationSystem reservationSystemInt;
    private final Scanner scanner;
    private static final String CUSTOMER_DATA_FILE = "customers.json";

    public MenuManager(ReservationSystem reservationSystemInt) {
        this.reservationSystemInt = reservationSystemInt;
        this.scanner = new Scanner(System.in);
    }

    public void displayMenu() {
        int choice;
        do {
            System.out.println("Airline Reservation Simulator");
            System.out.println("1. Create Customer");
            System.out.println("2. Create Flight");
            System.out.println("3. Book Ticket");
            System.out.println("4. Add Offer");
            System.out.println("5. View Ticket History");
            System.out.println("6. Add Baggage");
            System.out.println("7. Change Flight Date");
            System.out.println("8. Update Ticket Status");
            System.out.println("9. View Offers");
            System.out.println("10. FAQ");
            System.out.println("0. Exit");
            System.out.print("Enter your choice: ");
            choice = getIntInput();

            switch (choice) {
                case 1:
                    createCustomer();
                    break;
                case 2:
                    createFlight();
                    break;
                case 3:
                    bookTicket();
                    break;
                case 4:
                    addOffer();
                    break;
                case 5:
                    viewTicketHistory();
                    break;
                case 6:
                    addBaggage();
                    break;
                case 7:
                    changeFlightDate();
                    break;
                case 8:
                    updateTicketStatus();
                    break;
                case 9:
                    viewOffers();
                    break;
                case 10:
                    displayFAQ();
                    break;
                case 0:
                    System.out.println("Exiting...");
                    break;
                default:
                    System.out.println("Invalid choice! Please try again.");
            }
        } while (choice != 0);

        scanner.close();
    }

    private void createCustomer() {
        System.out.print("Enter customer name: ");
        String name = scanner.nextLine();
        System.out.print("Enter customer email: ");
        String email = scanner.nextLine();

        Customer customer = reservationSystemInt.getCustomerManager().createCustomer(name, email);
        System.out.println("Customer created successfully! Customer ID: " + customer.getCustomerId());

        // Save customers to JSON file
        List<Customer> customers = reservationSystemInt.getCustomerManager().getAllCustomers();
        CustomerDataManager.saveCustomers(customers);
    }

    private void createFlight() {
        System.out.print("Enter origin: ");
        String origin = scanner.nextLine();
        System.out.print("Enter destination: ");
        String destination = scanner.nextLine();
        System.out.print("Enter airline: ");
        String airline = scanner.nextLine();
        System.out.print("Enter departure date (yyyy-MM-dd): ");
        String depDateStr = scanner.nextLine();
        System.out.print("Enter arrival date (yyyy-MM-dd): ");
        String arrDateStr = scanner.nextLine();
        try {
            Date depDate = new SimpleDateFormat("yyyy-MM-dd").parse(depDateStr);
            Date arrDate = new SimpleDateFormat("yyyy-MM-dd").parse(arrDateStr);
            Flight flight = reservationSystemInt.getFlightManager().createFlight(origin, destination, airline, depDate, arrDate);
            System.out.println("Flight created successfully! Flight ID: " + flight.getFlightId());
        } catch (ParseException e) {
            System.out.println("Invalid date format. Please try again.");
        }
    }

    private void bookTicket() {
        System.out.print("Enter customer ID: ");
        int customerId = getIntInput();
        System.out.print("Enter flight ID: ");
        int flightId = getIntInput();
        System.out.print("Enter ticket class (ECONOMY/BUSINESS): ");
        String ticketClassStr = scanner.nextLine();
        System.out.print("Enter number of people: ");
        int numberOfPeople = getIntInput();

        try {
            TicketClass ticketClass = TicketClass.valueOf(ticketClassStr.toUpperCase());
            Customer customer = reservationSystemInt.getCustomerManager().getCustomer(customerId);
            Flight flight = reservationSystemInt.getFlightManager().getFlight(flightId);
            Ticket ticket = reservationSystemInt.getTicketManager().bookTicket(customer, flight, ticketClass, numberOfPeople);
            System.out.println("Ticket booked successfully!");

            // Calculate total price
            double totalPrice = ticketClass.getPrice() * numberOfPeople;

            // Proceed to payment
            System.out.print("Enter payment method (CREDIT_CARD, DEBIT_CARD, PAYPAL): ");
            String paymentMethodStr = scanner.nextLine();

            PaymentMethods paymentMethod = PaymentMethods.valueOf(paymentMethodStr.toUpperCase());
            reservationSystemInt.getPaymentManager().makePayment(paymentMethod, totalPrice);
            System.out.println("Payment made successfully!");

            List<Customer> customers = reservationSystemInt.getCustomerManager().getAllCustomers();
            JsonUtils.saveCustomersToFile(customers, CUSTOMER_DATA_FILE);

        } catch (IndexOutOfBoundsException | IllegalArgumentException e) {
            System.out.println("Invalid input. Please try again.");
        }
    }

    private void addOffer() {
        System.out.print("Enter offer details: ");
        String offer = scanner.nextLine();
        reservationSystemInt.getOfferManager().addOffer(offer);
        System.out.println("Offer added successfully!");
    }

    private void viewTicketHistory() {
        System.out.print("Enter customer ID: ");
        int customerId = getIntInput();
        try {
            Customer customer = reservationSystemInt.getCustomerManager().getCustomer(customerId);
            if (customer != null) {
                List<Ticket> tickets = reservationSystemInt.getTicketManager().getTicketsByCustomer(customer);

                if (!tickets.isEmpty()) {
                    System.out.println("Ticket History for " + customer.getName() + ":");
                    for (Ticket ticket : tickets) {
                        Flight flight = reservationSystemInt.getFlightManager().getFlight(ticket.getFlightId());
                        if (flight != null) {
                            System.out.println("- Ticket ID: " + ticket.getTicketId()
                                    + ", Flight ID: " + flight.getFlightId()
                                    + ", Origin: " + flight.getOrigin()
                                    + ", Destination: " + flight.getDestination()
                                    + ", Airline: " + flight.getAirline()
                                    + ", Departure Date: " + flight.getDepartureDate()
                                    + ", Arrival Date: " + flight.getArrivalDate()
                                    + ", Class: " + ticket.getTicketClass()
                                    + ", Status: " + ticket.getStatus());
                        } else {
                            System.out.println("- Ticket ID: " + ticket.getTicketId()
                                    + ", Flight information not available.");
                        }
                    }
                } else {
                    System.out.println("No tickets found for customer " + customer.getName());
                }
            } else {
                System.out.println("Invalid customer ID. Please try again.");
            }
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Error accessing ticket information. Please try again.");
        }
    }

    private void addBaggage() {
        System.out.print("Enter ticket ID: ");
        int ticketId = getIntInput();
        System.out.print("Enter number of carry-on baggage: ");
        int carryOn = getIntInput();
        System.out.print("Enter number of checked baggage: ");
        int checked = getIntInput();
        try {
            Ticket ticket = reservationSystemInt.getTicketManager().getTicket(ticketId);
            reservationSystemInt.getTicketManager().addBaggage(ticket, carryOn, checked);
            System.out.println("Baggage added successfully!");
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Invalid ticket ID. Please try again.");
        }
    }

    private void changeFlightDate() {
        System.out.print("Enter ticket ID: ");
        int ticketId = getIntInput();
        scanner.nextLine();
        System.out.print("Enter new departure date (yyyy-MM-dd): ");
        String newDepDateStr = scanner.nextLine();
        System.out.print("Enter new arrival date (yyyy-MM-dd): ");
        String newArrDateStr = scanner.nextLine();
        try {
            Date newDepDate = new SimpleDateFormat("yyyy-MM-dd").parse(newDepDateStr);
            Date newArrDate = new SimpleDateFormat("yyyy-MM-dd").parse(newArrDateStr);
            Ticket ticket = reservationSystemInt.getTicketManager().getTicket(ticketId);
            reservationSystemInt.getTicketManager().changeFlightDate(ticket, newDepDate, newArrDate);
            System.out.println("Flight date changed successfully!");
        } catch (ParseException e) {
            System.out.println("Invalid date format. Please try again.");
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Invalid ticket ID. Please try again.");
        }
    }

    private void updateTicketStatus() {
        System.out.print("Enter ticket ID: ");
        int ticketId = getIntInput();
        scanner.nextLine();
        System.out.print("Enter new status: ");
        String status = scanner.nextLine();
        try {
            Ticket ticket = reservationSystemInt.getTicketManager().getTicket(ticketId);
            reservationSystemInt.getTicketManager().updateTicketStatus(ticket, status);
            System.out.println("Ticket status updated successfully!");
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Invalid ticket ID. Please try again.");
        }
    }

    private void viewOffers() {
        reservationSystemInt.getOfferManager().listOffers();
    }

    private void displayFAQ() {
        reservationSystemInt.getFaqManager().displayFAQ();
    }

    private int getIntInput() {
        while (!scanner.hasNextInt()) {
            System.out.print("Invalid input. Please enter a number: ");
            scanner.next();
        }
        int input = scanner.nextInt();
        scanner.nextLine(); // Consume newline
        return input;
    }
}

File: hw/gomez/u2/hw15MongoDBGraphicInterfaces/MenuManager.py
import tkinter as tk
from tkinter import messagebox, simpledialog
from pymongo import MongoClient
from datetime import datetime

# Conectar a MongoDB (reemplaza con tu URI de conexión)
client = MongoClient("mongodb+srv://JoffreGQ:joffre2004@cluster0.yztxrwi.mongodb.net/")
db = client.airline_reservation_system
customers_collection = db.customers
flights_collection = db.flights
tickets_collection = db.tickets

class AirlineReservationGUI:
    def __init__(self, root):
        self.root = root
        self.root.title("Airline Reservation System")

        # Combo box para las opciones del menú
        self.menu_options = [
            "Create Customer",
            "Create Flight",
            "Book Ticket",
            "Add Offer",
            "View Ticket History",
            "Add Baggage",
            "Change Flight Date",
            "Update Ticket Status",
            "View Offers",
            "FAQ",
            "Exit"
        ]
        self.selected_option = tk.StringVar()
        self.selected_option.set(self.menu_options[0])  # Opción por defecto

        # Crear el menú desplegable
        tk.Label(self.root, text="Airline Reservation Simulator").pack()
        tk.Label(self.root, text="Select an option:").pack()
        self.menu_dropdown = tk.OptionMenu(self.root, self.selected_option, *self.menu_options)
        self.menu_dropdown.pack(pady=10)

        # Botón para ejecutar la opción seleccionada
        tk.Button(self.root, text="Execute", command=self.execute_option).pack(pady=10)

    def execute_option(self):
        selected = self.selected_option.get()

        if selected == "Exit":
            self.root.destroy()
        else:
            # Ejecutar la función correspondiente según la opción seleccionada
            if selected == "Create Customer":
                self.create_customer()
            elif selected == "Create Flight":
                self.create_flight()
            elif selected == "Book Ticket":
                self.book_ticket()
            elif selected == "Add Offer":
                self.add_offer()
            elif selected == "View Ticket History":
                self.view_ticket_history()
            elif selected == "Add Baggage":
                self.add_baggage()
            elif selected == "Change Flight Date":
                self.change_flight_date()
            elif selected == "Update Ticket Status":
                self.update_ticket_status()
            elif selected == "View Offers":
                self.view_offers()
            elif selected == "FAQ":
                self.display_faq()

    def create_customer(self):
        name = simpledialog.askstring("Enter customer name", "Customer Name:")
        email = simpledialog.askstring("Enter customer email", "Customer Email:")

        if name and email:
            customer_data = {
                "name": name,
                "email": email,
                "created_at": datetime.now()
            }
            result = customers_collection.insert_one(customer_data)
            if result.inserted_id:
                messagebox.showinfo("Success", "Customer created successfully!")
            else:
                messagebox.showerror("Error", "Failed to create customer.")
        else:
            messagebox.showwarning("Warning", "Please enter both name and email.")

    def create_flight(self):
        origin = simpledialog.askstring("Enter origin", "Origin:")
        destination = simpledialog.askstring("Enter destination", "Destination:")
        airline = simpledialog.askstring("Enter airline", "Airline:")
        dep_date_str = simpledialog.askstring("Enter departure date", "Departure Date (yyyy-MM-dd):")
        arr_date_str = simpledialog.askstring("Enter arrival date", "Arrival Date (yyyy-MM-dd):")

        try:
            dep_date = datetime.strptime(dep_date_str, "%Y-%m-%d")
            arr_date = datetime.strptime(arr_date_str, "%Y-%m-%d")
            flight_data = {
                "origin": origin,
                "destination": destination,
                "airline": airline,
                "departure_date": dep_date,
                "arrival_date": arr_date,
                "created_at": datetime.now()
            }
            result = flights_collection.insert_one(flight_data)
            if result.inserted_id:
                messagebox.showinfo("Success", "Flight created successfully!")
            else:
                messagebox.showerror("Error", "Failed to create flight.")
        except ValueError:
            messagebox.showerror("Error", "Invalid date format. Please use yyyy-MM-dd.")

    def book_ticket(self):
        customer_id = simpledialog.askinteger("Enter customer ID", "Customer ID:")
        flight_id = simpledialog.askinteger("Enter flight ID", "Flight ID:")
        ticket_class_str = simpledialog.askstring("Enter ticket class", "Ticket Class (ECONOMY/BUSINESS):")
        number_of_people = simpledialog.askinteger("Enter number of people", "Number of People:")

        try:
            # Validar que el customer_id y flight_id existen en MongoDB (no implementado aquí por simplicidad)

            ticket_data = {
                "customer_id": customer_id,
                "flight_id": flight_id,
                "ticket_class": ticket_class_str,
                "number_of_people": number_of_people,
                "status": "BOOKED",
                "created_at": datetime.now()
            }
            result = tickets_collection.insert_one(ticket_data)
            if result.inserted_id:
                messagebox.showinfo("Success", "Ticket booked successfully!")
            else:
                messagebox.showerror("Error", "Failed to book ticket.")
        except Exception as e:
            messagebox.showerror("Error", f"Failed to book ticket: {str(e)}")

    def add_offer(self):
        offer = simpledialog.askstring("Enter offer details", "Offer Details:")

        if offer:
            # Guardar la oferta en MongoDB (implementa según tu estructura de datos)
            messagebox.showinfo("Success", "Offer added successfully!")
        else:
            messagebox.showwarning("Warning", "Please enter offer details.")

    def view_ticket_history(self):
        customer_id = simpledialog.askinteger("Enter customer ID", "Customer ID:")

        try:
            # Obtener historial de tickets del cliente desde MongoDB (no implementado aquí por simplicidad)

            # Mostrar historial de tickets (implementa según tu estructura de datos)
            messagebox.showinfo("Ticket History", "Implementación pendiente")
        except Exception as e:
            messagebox.showerror("Error", f"Failed to retrieve ticket history: {str(e)}")

    def add_baggage(self):
        ticket_id = simpledialog.askinteger("Enter ticket ID", "Ticket ID:")
        carry_on = simpledialog.askinteger("Enter number of carry-on baggage", "Carry-on Baggage:")
        checked = simpledialog.askinteger("Enter number of checked baggage", "Checked Baggage:")

        try:
            # Validar que el ticket_id existe en MongoDB (no implementado aquí por simplicidad)

            # Agregar equipaje al ticket (implementa según tu estructura de datos)
            messagebox.showinfo("Success", "Baggage added successfully!")
        except Exception as e:
            messagebox.showerror("Error", f"Failed to add baggage: {str(e)}")

    def change_flight_date(self):
        ticket_id = simpledialog.askinteger("Enter ticket ID", "Ticket ID:")
        new_dep_date_str = simpledialog.askstring("Enter new departure date", "New Departure Date (yyyy-MM-dd):")
        new_arr_date_str = simpledialog.askstring("Enter new arrival date", "New Arrival Date (yyyy-MM-dd):")

        try:
            new_dep_date = datetime.strptime(new_dep_date_str, "%Y-%m-%d")
            new_arr_date = datetime.strptime(new_arr_date_str, "%Y-%m-%d")

            # Actualizar fechas del vuelo en MongoDB (implementa según tu estructura de datos)
            messagebox.showinfo("Success", "Flight dates changed successfully!")
        except ValueError:
            messagebox.showerror("Error", "Invalid date format. Please use yyyy-MM-dd.")

    def update_ticket_status(self):
        ticket_id = simpledialog.askinteger("Enter ticket ID", "Ticket ID:")
        new_status = simpledialog.askstring("Enter new status", "New Status:")

        try:
            # Actualizar estado del ticket en MongoDB (implementa según tu estructura de datos)
            messagebox.showinfo("Success", "Ticket status updated successfully!")
        except Exception as e:
            messagebox.showerror("Error", f"Failed to update ticket status: {str(e)}")

    def view_offers(self):
        # Implementar función para ver ofertas (consulta MongoDB y muestra en una ventana)
        messagebox.showinfo("Offers", "Implementación pendiente")

    def display_faq(self):
        # Implementar función para mostrar FAQ (consulta MongoDB y muestra en una ventana)
        messagebox.showinfo("FAQ", "Implementación pendiente")

if __name__ == "__main__":
    root = tk.Tk()
    app = AirlineReservationGUI(root)
    root.mainloop()
